src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +64,287 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
+															   sizeFunctionType);
+
+		connection = GetNodeConnection(connectionFlag, workerNodeName, workerNodePort);
+
+		executeCommand = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+													  &queryResult);
+
+		/*
+		 * If there is a connection error and the table is not reference table,
+		 * close the connection and quit. If it is a reference table continue
+		 * to get the size from another node.
+		 */
+		if (executeCommand != 0)
+		{
+			PQclear(queryResult);
+
+			ForgetResults(connection);
+
+			if (distributedTableCacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
+			{
+				continue;
+			}
+
+			ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+							errmsg("cannot get the size")));
+		}
+
+		tableSizeString = PQgetvalue(queryResult, 0, 0);
+		totalRelationSize += atoi(tableSizeString);
+
+		PQclear(queryResult);
+
+		ForgetResults(connection);
+
+		/* size of single shard is enough to find the size of reference table */
+		if (distributedTableCacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
+		{
+			return totalRelationSize * list_length(workerNodeList);
+		}
+	}
+
+	/* size must be returned inside the loop for reference tables */
+	if (distributedTableCacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size")));
+	}
+
+	return totalRelationSize;
+}
+
+
+/*
+ * citus_total_relation_size accepts a table name and
+ * returns a distributed table and its indexes' total relation size.
+ */
+Datum
+citus_total_relation_size(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint64 totalRelationSize = 0;
+
+	totalRelationSize = SupportFunctionForCitusSize(relationId,
+													CITUS_TOTAL_RELATION_SIZE);
+
+	PG_RETURN_INT64(totalRelationSize);
+}
+
+
+/*
+ * citus_table_size accepts a table name and
+ * returns a distributed table's total relation size.
+ *  */
+Datum
+citus_table_size(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint64 totalRelationSize = 0;
+
+	totalRelationSize = SupportFunctionForCitusSize(relationId, CITUS_TABLE_SIZE);
+
+	PG_RETURN_INT64(totalRelationSize);
+}
+
+
+/*
+ * citus_relation_size accept a table name and
+ * returns a relation's 'main' fork's size.
+ */
+Datum
+citus_relation_size(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint64 totalRelationSize = 0;
+
+	totalRelationSize = SupportFunctionForCitusSize(relationId, CITUS_RELATION_SIZE);
+
+	PG_RETURN_INT64(totalRelationSize);
+}
+
+
+/*
+ * ErrorIfNotSuitableToGetSizeetermine determines whether the table is
+ * suitable to find its' size with internal functions.
+ *
+ */
+void
+ErrorIfNotSuitableToGetSize(Oid relationId)
+{
+	char *relationName = get_rel_name(relationId);
+	DistTableCacheEntry *distributedTableCacheEntry = NULL;
+
+	if (!IsDistributedTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+						errmsg("relation %s is not distributed",
```

Comment:
Let's say errmsg("cannot get the size") and errdetail("relation %s is not distributed", relationName)